connectors/sources/jira.py

```diff
@@ -0,0 +1,496 @@
+
+        async for response in self._api_call(url_name=PROJECT):
+            response = await response.json()
+            for project in response:
+                yield {
+                    "_id": f"{project['name']}-{project['id']}",
+                    "_timestamp": iso_utc(
+                        when=datetime.now(pytz.timezone(self.timezone))
+                    ),
+                    "Type": "Project",
+                    "Project": project,
+                }
+
+    async def _get_issues(self):
+        """Get issues with the help of REST APIs
+
+        Yields:
+            Dictionary: Jira issue to get indexed
+            issue (dict): Issue response to fetch the attachments
+        """
+        async for response in self._paginated_api_call(url_name=ISSUES):
+            for issue in response.get("issues", []):
+                async for response in self._api_call(
+                    url_name=ISSUE_DATA, id=issue["key"]
+                ):
+                    issue = await response.json()
+                    if issue:
+                        response_fields = issue.get("fields")
+                        yield {
+                            "_id": f"{response_fields['project']['name']}-{issue['key']}",
+                            "_timestamp": response_fields["updated"],
+                            "Type": response_fields["issuetype"]["name"],
+                            "Issue": response_fields,
+                        }, issue
+
+    async def _get_attachments(self, attachments, issue_key):
+        """Get attachments of a specific issue
+
+        Args:
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+
+        Yields:
+            Dictionary: Jira attachment on the given issue
+            attachment (dict): Attachment dictionary for extracting the content
+        """
+        for attachment in attachments:
+            yield {
+                "_id": f"{issue_key}-{attachment['id']}",
+                "title": attachment["filename"],
+                "type": "Attachment",
+                "issue": issue_key,
+                "_timestamp": attachment["created"],
+                "size": attachment["size"],
+            }, attachment
+
+    async def _grab_content(self, attachments, issue_key):
+        """Coroutine to add attachments to Queue and get content
+
+        Args:
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+        """
+        async for content, attachment in self._get_attachments(
+            attachments=attachments, issue_key=issue_key
+        ):
+            await self.queue.put(
+                (  # pyright: ignore
+                    content,
+                    partial(
+                        self.get_content,
+                        issue_key=issue_key,
+                        attachment=copy(attachment),
```

Comment:
why `copy` here?

Reply:
Here, attachment is a dictionary so any changes in the `partial` should not be reflected in the below code. Hence, we passed the `copy` to avoid changes done by the partial in the passed param.